authorization/role/repository/default_role_mapping.go

```diff
@@ -0,0 +1,202 @@
+}
+
+// TableName overrides the table name settings in Gorm to force a specific table name
+// in the database.
+func (m *GormDefaultRoleMappingRepository) TableName() string {
+	return "default_role_mapping"
+}
+
+// CheckExists returns nil if the given ID exists otherwise returns an error
+func (m *GormDefaultRoleMappingRepository) CheckExists(ctx context.Context, ID uuid.UUID) (bool, error) {
+	defer goa.MeasureSince([]string{"goa", "db", "default_role_mapping", "exists"}, time.Now())
+
+	var exists bool
+	query := fmt.Sprintf(`
+		SELECT EXISTS (
+			SELECT 1 FROM %[1]s
+			WHERE
+				default_role_mapping_id=$1
+				AND deleted_at IS NULL
+		)`, m.TableName())
+
+	err := m.db.CommonDB().QueryRow(query, ID.String()).Scan(&exists)
+	if err == nil && !exists {
+		return exists, errors.NewNotFoundError(m.TableName(), ID.String())
+	}
+	if err != nil {
+		return false, errors.NewInternalError(ctx, errs.Wrapf(err, "unable to verify if %s exists", m.TableName()))
+	}
+	return exists, nil
+}
+
+// CRUD Functions
+
+// Load returns a single RoleMapping as a Database Model
+func (m *GormDefaultRoleMappingRepository) Load(ctx context.Context, id uuid.UUID) (*DefaultRoleMapping, error) {
+	defer goa.MeasureSince([]string{"goa", "db", "default_role_mapping", "load"}, time.Now())
+	var native DefaultRoleMapping
+	err := m.db.Table(m.TableName()).
+		Preload("ResourceType").
+		Preload("FromRole").
+		Preload("ToRole").
+		Where("default_role_mapping_id = ?", id).Find(&native).Error
+	if err == gorm.ErrRecordNotFound {
+		return nil, errors.NewNotFoundError("default_role_mapping", id.String())
+	}
+	return &native, errs.WithStack(err)
+}
+
+// Create creates a new record.
+func (m *GormDefaultRoleMappingRepository) Create(ctx context.Context, u *DefaultRoleMapping) error {
+	defer goa.MeasureSince([]string{"goa", "db", "default_role_mapping", "create"}, time.Now())
+	if u.DefaultRoleMappingID == uuid.Nil {
+		u.DefaultRoleMappingID = uuid.NewV4()
+	}
+	err := m.db.Create(u).Error
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"default_role_mapping_id": u.DefaultRoleMappingID,
+			"err": err,
+		}, "unable to create the default role mapping")
+		return errs.WithStack(err)
+	}
+	log.Debug(ctx, map[string]interface{}{
+		"default_role_mapping_id": u.DefaultRoleMappingID,
+	}, "Default role mapping created!")
+	return nil
+}
+
+// Save modifies a single record
+func (m *GormDefaultRoleMappingRepository) Save(ctx context.Context, model *DefaultRoleMapping) error {
+	defer goa.MeasureSince([]string{"goa", "db", "default_role_mapping", "save"}, time.Now())
+
+	obj, err := m.Load(ctx, model.DefaultRoleMappingID)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"default_role_mapping_id": model.DefaultRoleMappingID,
+			"err": err,
+		}, "unable to update default role mapping")
+		return errs.WithStack(err)
+	}
+	err = m.db.Model(obj).Select("ResourceTypeID", "FromRoleID", "ToRoleID").Updates(model).Error
```

Comment:
We use .Updates() instead of Save() almost everywhere but in Identity and User repos. Main difference is that Update doesn't update empty fields. If the field of "model" is empty then GORM treats it's as not modified and doesn't update it. Which is not critical for default mappings but still..

Save() updates all the fields. So, as result it's impossible to set some field to an empty value (for example an empty string) using Updates().

Updates also require Load() first, so it's two DB calls instead of one in case of Save().
See https://github.com/fabric8-services/fabric8-auth/blob/master/account/repository/identity.go#L232 as an example.

I think we should start using Save() everywhere and use Updates() only if we really need it. WDYT?

Reply:
I've changed this to save() instead.  In fact, I had some issues using Updates() as it was trying to update the related role record also for some reason.

Reply:
Looks better but you don't have to load the object before updating it.